### ISSUE TITLE
QE crashes when toasted data is dispatched

### ISSUE THREAD

**Issue by kuien** [greenplum-db/gpdb#661]
When deploying PostGIS/Raster onto gpdb master, we met this issue. The case is:
1. Given a **variable-length UDT** declared **without** recv/send functions
2. Create a table with a UDT column
3. Insert a row with UDT instance size > CHUNKSZ (e.g., 8K) as the first line

You may catch below error:

``` sql
test=# INSERT INTO comment_tbl_out VALUES('10000*e', fillCommentWithLen((SELECT comment FROM comment_tbl WHERE id='e'), 10000 ));
ERROR:  invalid memory alloc request size 1701143913 (context 'TupSerMemCtxt') (tuptoaster.c:338) (mcxt.c:477)  (seg1 slice1 r7:40001 pid=11672) (cdbdisp.c:1324)
CONTEXT:  SQL statement "select Ta.id, Ta.comment from public.comment_tbl_out as Ta  limit 7500 "
```

But, if I insert lines with size from 1B to 100MB, that is, the first few lines needs not to be toasted, it is OK.

``` sql
     id      |  length
-------------+-----------
 1*a         |         1
 10*b        |        10
 100*c       |       100
 1000*d      |      1000
 10000*e     |     10000
 100000*f    |    100000
 1000000*g   |   1000000
 10000000*h  |  10000000
 100000000*i | 100000000
(9 rows)
```

However, below SQL will make QE crashed too.

``` sql
select Ta.id, Ta.comment from public.comment_tbl_out as Ta  limit 7500;
```

To re-produce this issue, we write an gpdb extension in attached file.
[udt.zip](https://github.com/greenplum-db/gpdb/files/229517/udt.zip)

``` sh
    In this extension, you can run test in following steps:
      ./run.sh       # anything is fine
      ./run_bug.sh   # QE crashes

    The common parts of run.sh and run_bug.sh are:
      make
      make install
      createdb test

    The different part is:
      psql test -f pg_comment(_bug).sql
      psql test -f test(_bug).sql 

    Author: Kuien Liu, Haozhou Wang
```

Besides, Luis Macedo also reports potentially similar issue, I paste his message from slack here

```
Luis Macedo [9:08 PM] 
@kuien: My customer that is having issues with parsing XML on ETL phase has the same error message on their logs... Any chance is the same bug?
```

**Comment by foyzur** [comment]
Query execution team is taking a look at this.

**Comment by foyzur** [comment]
## RCA

The motion node uses a binary serializer to serialize a type inside SerializeTupleIntoChunks. This binary serializer is supposed to produce at varlena with header to indicate compression info, toasting info etc. The serializer is passed into SerializeTupleIntoChunks as pSerInfo. This pSerInfo originally came during motion layer setup from InitSerTupInfo, which uses CAQL to read the type functions from pg_type.

The InitSerTupInfo forcefully populates out/in functions in send/recv function if send/recv are not set:

```
			/* If we don't have both binary routines */
			if (!OidIsValid(pt->typsend) || !OidIsValid(pt->typreceive))
			{
				/* Use the normal text routines (slower) */
				if (!OidIsValid(pt->typoutput))
					ereport(ERROR,
						(errcode(ERRCODE_UNDEFINED_FUNCTION),
						 errmsg("no output function available for type %s",
								format_type_be(attrInfo->atttypid))));
				if (!OidIsValid(pt->typinput))
					ereport(ERROR,
						(errcode(ERRCODE_UNDEFINED_FUNCTION),
						 errmsg("no input function available for type %s",
								format_type_be(attrInfo->atttypid))));
		
				attrInfo->typsend = pt->typoutput;
				attrInfo->send_typio_param = getTypeIOParam(typeTuple);
				attrInfo->typisvarlena = (!pt->typbyval) && (pt->typlen == -1);
				attrInfo->typrecv = pt->typinput;
				attrInfo->recv_typio_param = getTypeIOParam(typeTuple);
			}

```

The problem is, the out function in our case returns a string, not a varlena. So, the headers are not set. Therefore, during byte serialization we will actually obtain a string datum. This datum is then passed to DatumGetByteaP to do the detoasting (if necessary, of course). Unfortunately, pg_detoast_datum cannot find the header info (that was supposed to be in varlena header) as we only have a string datum, and not a varlena. Hence, it obtains a wrong size information.

**Comment by foyzur** [comment]
While in our reference guide the Send and Receive functions are optional, this seems to be a bad idea as they are necessary for our MPP database. The "optional" nature of send/receive functions most likely came from postgresql. From postgresql doc:

```
The optional receive_function converts the type's external binary representation to the internal representation. If this function is not supplied, the type cannot participate in binary input. The binary representation should be chosen to be cheap to convert to internal form, while being reasonably portable. (For example, the standard integer data types use network byte order as the external binary representation, while the internal representation is in the machine's native byte order.) The receive function should perform adequate checking to ensure that the value is valid. The receive function can be declared as taking one argument of type internal, or as taking three arguments of types internal, oid, integer. The first argument is a pointer to a StringInfo buffer holding the received byte string; the optional arguments are the same as for the text input function. The receive function must return a value of the data type itself. Usually, a receive function should be declared STRICT; if it is not, it will be called with a NULL first parameter when reading a NULL input value. The function must still return NULL in this case, unless it raises an error. (This case is mainly meant to support domain receive functions, which might need to reject NULL inputs.) Similarly, the optional send_function converts from the internal representation to the external binary representation. If this function is not supplied, the type cannot participate in binary output. The send function must be declared as taking one argument of the new data type. The send function must return type bytea. Send functions are not invoked for NULL values.
```

But, for our use case we must provide these during motion serialization/deserialization. Therefore, it is not optional for us. Moreover, falling back to "out" function for "send" is a terrible idea as they don't even have the same return type. There could be two solutions:

1. Disable optional nature in parser
2. Error out during execution if send/receive are not valid.

**Comment by danielgustafsson** [comment]
In this case I would prefer to error out early and disable in the parser.

Thanks for the good writeup!

**Comment by foyzur** [comment]
Hi @danielgustafsson, I absolutely agree with you that the solution should be in parser. Unfortunately, adding it in parser would be a more disruptive fix for our customers as the problem only manifests if we have a motion layer serialization/deserialization requirement. Therefore, I am adding the fix to executor, only checking when motion initializes the sed/receive function. I will open a PR soon. If you feel strongly that it should be in the parser, please comment on the PR and we can do that too.

**Comment by danielgustafsson** [comment]
Added my $0.02 on the PR, thanks for working on this issue!

**Comment by hlinnaka** [comment]
This was finally fixed by PR #2906. Closing, thanks for the report!
